File: client/src/main/java/ir/asparsa/hobbytaste/ui/list/holder/GalleryViewHolder.java
package ir.asparsa.hobbytaste.ui.list.holder;

import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import butterknife.BindView;
import butterknife.ButterKnife;
import com.squareup.picasso.Callback;
import com.squareup.picasso.Picasso;
import ir.asparsa.android.core.logger.L;
import ir.asparsa.android.ui.fragment.recycler.BaseRecyclerFragment;
import ir.asparsa.android.ui.list.holder.BaseViewHolder;
import ir.asparsa.hobbytaste.BuildConfig;
import ir.asparsa.hobbytaste.R;
import ir.asparsa.hobbytaste.database.model.BannerModel;
import ir.asparsa.hobbytaste.ui.list.HorizontalSpaceItemDecoration;
import ir.asparsa.hobbytaste.ui.list.data.GalleryData;
import rx.Observer;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by hadi on 12/12/2016 AD.
 */
public class GalleryViewHolder extends BaseViewHolder<GalleryData> {

    private final Adapter mAdapter;

    @BindView(R.id.gallery_recycler)
    RecyclerView mRecyclerView;

    public GalleryViewHolder(
            View itemView,
            Observer<BaseRecyclerFragment.Event> observer,
            Bundle savedInstanceState
    ) {
        super(itemView, observer, savedInstanceState);
        ButterKnife.bind(this, itemView);

        LinearLayoutManager layoutManager = new LinearLayoutManager(
                itemView.getContext(), LinearLayoutManager.HORIZONTAL, false);

        mAdapter = new Adapter();
        mRecyclerView.setLayoutManager(layoutManager);
        mRecyclerView.setAdapter(mAdapter);

        mRecyclerView.addItemDecoration(new HorizontalSpaceItemDecoration(
                (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 10,
                                                itemView.getContext().getResources().getDisplayMetrics()),
                (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 2,
                                                itemView.getContext().getResources().getDisplayMetrics())));
    }

    @Override
    public void onBindView(GalleryData data) {
        mAdapter.banners = data.getBanners();
        mAdapter.notifyDataSetChanged();
    }

    private class Adapter extends RecyclerView.Adapter {

        private List<BannerModel> banners = new ArrayList<>();

        @Override
        public RecyclerView.ViewHolder onCreateViewHolder(
                ViewGroup parent,
                int viewType
        ) {
            return new ViewHolder(LayoutInflater.from(parent.getContext()).inflate(viewType, parent, false));
        }

        @Override
        public void onBindViewHolder(
                RecyclerView.ViewHolder holder,
                int position
        ) {
            if (holder instanceof ViewHolder) {
                ((ViewHolder) holder).onBindView(banners.get(position));
            }
        }

        @Override
        public int getItemCount() {
            return banners.size();
        }

        @Override
        public int getItemViewType(int position) {
            return R.layout.shot;
        }
    }

    class ViewHolder extends RecyclerView.ViewHolder implements Callback {

        @BindView(R.id.shot)
        ImageView mScreenshot;
        @BindView(R.id.try_again)
        ImageView mTryAgain;
        @BindView(R.id.progress_bar)
        ProgressBar mProgressBar;

        private String urlString;

        ViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }

        void onBindView(final BannerModel banner) {
            boolean imageSet = false;
            L.i(getClass(), "mainUrl: " + banner.getMainUrl() + ", thumbnailUrl: " + banner.getThumbnailUrl());
            if (TextUtils.isEmpty(banner.getThumbnailUrl())) {
                if (TextUtils.isEmpty(banner.getMainUrl())) {
                    itemView.setVisibility(View.GONE);
                } else {
                    setupScreenshot(banner.getMainUrl());
                    imageSet = true;
                }
            } else {
                setupScreenshot(banner.getThumbnailUrl());
                imageSet = true;
            }

            if (imageSet) {
                mScreenshot.setOnClickListener(new View.OnClickListener() {
                    @Override public void onClick(View view) {
                        if (mObserver != null) {
                            mObserver.onNext(new OnScreenshotClick(banner));
                        }
                    }
                });
            }
        }

        private void setupScreenshot(final String urlString) {
            this.urlString = urlString;
            final Picasso builder = Picasso.with(itemView.getContext());
            if (BuildConfig.DEBUG) {
                builder.setLoggingEnabled(true);
            }
            builder.load(urlString)
                   .into(mScreenshot, this);
        }

        @Override public void onSuccess() {
            mProgressBar.setVisibility(View.GONE);
            mTryAgain.setVisibility(View.GONE);
        }

        @Override public void onError() {
            mProgressBar.setVisibility(View.GONE);
            mTryAgain.setVisibility(View.VISIBLE);
            mTryAgain.setOnClickListener(new View.OnClickListener() {
                @Override public void onClick(View v) {
                    mProgressBar.setVisibility(View.VISIBLE);
                    mTryAgain.setVisibility(View.GONE);
                    setupScreenshot(urlString);
                }
            });
        }
    }

    public static class OnScreenshotClick implements BaseRecyclerFragment.Event {
        private BannerModel data;

        OnScreenshotClick(
                BannerModel data
        ) {
            this.data = data;
        }

        public BannerModel getData() {
            return data;
        }
    }

}

File: client/src/main/kotlin/ir/asparsa/hobbytaste/ui/list/holder/GalleryViewHolder.kt
package ir.asparsa.hobbytaste.ui.list.holder

import android.graphics.Bitmap
import android.graphics.PorterDuff
import android.graphics.drawable.Drawable
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.text.TextUtils
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.ProgressBar
import butterknife.ButterKnife
import com.github.kittinunf.reactiveandroid.view.rx_click
import com.squareup.picasso.Picasso
import com.squareup.picasso.Target
import ir.asparsa.android.core.logger.L
import ir.asparsa.android.ui.list.holder.BaseViewHolder
import ir.asparsa.hobbytaste.ApplicationLauncher
import ir.asparsa.hobbytaste.R
import ir.asparsa.hobbytaste.database.model.BannerModel
import ir.asparsa.hobbytaste.ui.list.HorizontalSpaceItemDecoration
import ir.asparsa.hobbytaste.ui.list.data.GalleryData
import kotterknife.bindView
import rx.Observable
import rx.subjects.PublishSubject
import rx.subjects.Subject
import javax.inject.Inject

/**
 * @author hadi
 */
class GalleryViewHolder(
        itemView: View
) : BaseViewHolder<GalleryData>(itemView) {

    @Inject
    lateinit var mPicasso: Picasso

    private val mAdapter: Adapter

    val mRecyclerView: RecyclerView by bindView(R.id.gallery_recycler)

    init {
        ApplicationLauncher.mainComponent().inject(this)
        ButterKnife.bind(this, itemView)

        val layoutManager = LinearLayoutManager(
                itemView.context, LinearLayoutManager.HORIZONTAL, false)

        mAdapter = Adapter()
        mRecyclerView.layoutManager = layoutManager
        mRecyclerView.adapter = mAdapter

        mRecyclerView.addItemDecoration(HorizontalSpaceItemDecoration(
                TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 10f,
                        itemView.context.resources.displayMetrics).toInt(),
                TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 2f,
                        itemView.context.resources.displayMetrics).toInt()))
    }

    override fun onBindView(data: GalleryData) {
        super.onBindView(data)
        mAdapter.banners = data.banners
        mAdapter.notifyDataSetChanged()
    }

    private inner class Adapter : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

        var subject: Subject<BannerModel, BannerModel> = PublishSubject.create<BannerModel>()
        var banners: List<BannerModel> = ArrayList()

        override fun onCreateViewHolder(
                parent: ViewGroup,
                viewType: Int
        ): RecyclerView.ViewHolder {
            val holder = ViewHolder(LayoutInflater.from(parent.context).inflate(viewType, parent, false))
            holder.clickStream().subscribe(subject)
            return holder
        }

        override fun onBindViewHolder(
                holder: RecyclerView.ViewHolder,
                position: Int
        ) {
            (holder as? ViewHolder)?.onBindView(banners[position])
        }

        override fun getItemCount(): Int {
            return banners.size
        }

        override fun getItemViewType(position: Int): Int {
            return R.layout.shot
        }
    }

    internal inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), Target {

        val mShotLayout: FrameLayout by bindView(R.id.shot_layout)
        val mScreenshot: ImageView by bindView(R.id.shot)
        val mTryAgain: ImageView by bindView(R.id.try_again)
        val mProgressBar: ProgressBar by bindView(R.id.progress_bar)

        private var urlString: String? = null
        private var banner: BannerModel? = null

        init {
            ButterKnife.bind(this, itemView)

            mTryAgain.drawable.mutate()
                    .setColorFilter(itemView.resources.getColor(R.color.refresh), PorterDuff.Mode.SRC_IN)
        }

        fun onBindView(banner: BannerModel) {
            this.banner = banner
            L.i(javaClass, "mainUrl: " + banner.mainUrl + ", thumbnailUrl: " + banner.thumbnailUrl)
            if (TextUtils.isEmpty(banner.thumbnailUrl)) {
                if (TextUtils.isEmpty(banner.mainUrl)) {
                    itemView.visibility = View.GONE
                } else {
                    setupScreenshot(banner.mainUrl)
                }
            } else {
                setupScreenshot(banner.thumbnailUrl)
            }
        }

        private fun setupScreenshot(urlString: String) {
            this.urlString = urlString
            mPicasso.load(urlString)
                    .into(this)
        }

        override fun onPrepareLoad(placeHolderDrawable: Drawable?) {
        }

        override fun onBitmapFailed(errorDrawable: Drawable?) {
            mProgressBar.visibility = View.GONE
            mTryAgain.visibility = View.VISIBLE
            mTryAgain.setOnClickListener {
                mProgressBar.visibility = View.VISIBLE
                mTryAgain.visibility = View.GONE
                setupScreenshot(urlString!!)
            }
        }

        override fun onBitmapLoaded(bitmap: Bitmap?, from: Picasso.LoadedFrom?) {
            mProgressBar.visibility = View.GONE
            mTryAgain.visibility = View.GONE
            if (bitmap == null) {
                return
            }
            val params = mScreenshot.layoutParams
            params.height = mScreenshot.resources.getDimensionPixelSize(R.dimen.shot_height)
            params.width = (params.height.toFloat() * bitmap.width / bitmap.height).toInt()
            mScreenshot.layoutParams = params
            mScreenshot.setImageBitmap(bitmap)
        }

        fun clickStream(): Observable<BannerModel> {
            return mShotLayout.rx_click().map {
                banner
            }
        }
    }

    fun clickStream(): Observable<BannerModel> {
        return mAdapter.subject
    }
}
